sweep/sweeper.go

```diff
@@ -854,9 +855,81 @@ func (s *UtxoSweeper) removeExclusiveGroup(group uint64) {
 		s.signalAndRemove(&outpoint, Result{
 			Err: ErrExclusiveGroupSpend,
 		})
+
+		// Make sure we remove conflicting transactions from the wallet.
+		err := s.removeConflictingSweepTx(&outpoint)
+		if err != nil {
+			log.Warnf("unable to remove conflicting sweep tx from "+
+				"wallet for outpoint %v : %v", outpoint, err)
+		}
 	}
 }
 
+// removeConflictingSweepTx removes any unconfirmed sweep transaction from the
+// wallet which spends the passed outpoint. It is necessary to cleanup the
+// unmined tx store for wallets which use neutrino as a backend.
+// Most common case is when a channel is force closed and anchor cpfp txns are
+// created to bump the initial commitment transaction. In this case an achor
+// cpfp is broadcasted for up to 3 commitment transactions (local,
+// remote-dangling, remote). Using neutrino all of those transactions will be
+// accepted (the commitment tx will be different in all of those cases) and have
+// to be removed as soon as one of them confirmes (they do have the same
+// ExclusiveGroup).
+func (s *UtxoSweeper) removeConflictingSweepTx(outpoint *wire.OutPoint) error {
+	// We need to fetch all sweep txns to check whether they spend the given
+	// outpoint.
+	pastSweepHashes, err := s.cfg.Store.ListSweeps()
+	if err != nil {
+		return err
+	}
+
+	for _, sweepHash := range pastSweepHashes {
+		sweepTx, err := s.cfg.Wallet.FetchTx(sweepHash)
+		if err != nil {
+			return err
+		}
+
+		// Transaction wasn't found in the wallet, may have already
+		// been replaced/removed.
+		if sweepTx == nil {
+			// If it was removed, then we'll play it safe and mark
+			// it as no longer need to be rebroadcasted.
+			s.cfg.Wallet.CancelRebroadcast(sweepHash)
+			continue
+		}
+
+		// Check whether the inputs of the sweep transaction match the
+		// passed outpoint.
+		var isConflicting bool
+		for _, txIn := range sweepTx.TxIn {
+			if *outpoint == txIn.PreviousOutPoint {
+				isConflicting = true
+				break
+			}
+		}
+
+		// If it did, then we'll signal the wallet to remove all the
+		// transactions that are descendants of outputs created by the
+		// sweepTx and the sweepTx itself.
+		if isConflicting {
+			log.Debugf("Removing sweep txid=%v from wallet: %v",
```

Comment:
Should we be a bit more verbose here? Something along the lines of "Removing conflicting sweep txid=%v from wallet because %v confirmed: %v"?

Reply:
Sounds good, changed it a small way from your suggestion:

`"Removing conflicting sweep txid=%v from wallet because input %v confirmed: %v`